#!/bin/bash
# prepare-sbi-unlabeled 0.0.1
# Generated by dx-app-wizard.
#
# Basic execution pattern: Your app will run on a single machine from
# beginning to end.
#
# Your job's input variables (if any) will be loaded as environment
# variables before this script runs.  Any array inputs will be loaded
# as bash arrays.
#
# Any code outside of main() (or any entry point you may add) is
# ALWAYS executed, followed by running the entry point itself.
#
# See https://wiki.dnanexus.com/Developer-Portal for tutorials on how
# to modify this file.

main() {

    dx-download-all-inputs --parallel

    dx-docker pull artifacts/variationanalysis-app:${Image_Version} &>/dev/null

    # configure
    genome_basename=`basename /input/indexed_genome/*.bases | cut -d. -f1`
    echo "export SBI_GENOME=/input/indexed_genome/${genome_basename}" >> /input/configure.sh
    alignment_basename=`basename /input/alignment/*.entries | cut -d. -f1`
    echo "export GOBY_ALIGNMENT=/input/alignment/${alignment_basename}" >> /input/configure.sh
    echo "export GOBY_NUM_SLICES=1" >> /input/configure.sh
    # adjust num threads to match number of cores -1:
    cpus=`grep physical  /proc/cpuinfo |grep id|wc -l`
    cpus=`echo $(( cpus / 3 * 2  ))`
    basename="${alignment_basename}"
    echo "export SBI_NUM_THREADS=${cpus}" >> /input/configure.sh
    echo "export INCLUDE_INDELS='true'" >> /input/configure.sh
    echo "export REALIGN_AROUND_INDELS='false'" >> /input/configure.sh
    echo "export REF_SAMPLING_RATE='0.01'" >> /input/configure.sh
    echo "export OUTPUT_PREFIX=${basename}" >> /input/configure.sh
    echo "export DATASET=${basename}" >> /input/configure.sh
    echo "export GOBY_NUM_SLICES='50'" >> /input/configure.sh
    echo "export DO_CONCAT='true'" >> /input/configure.sh
   
    cat /input/configure.sh

    # Run generate-genotype-sets-0.02.sh
    dx-docker run \
        -v /input/:/input \
        -v /output/sbi:/output/sbi \
        artifacts/variationanalysis-app:${Image_Version} \
        bash -c "source ~/.bashrc; source /input/configure.sh; cd /output/sbi; generate-genotype-unlabeled.sh 20g \"/input/alignment/${alignment_basename}\" \"/input/indexed_genome/${genome_basename}\" "

    ls -lrt /output/sbi

    # Arrange dataset in output directory and upload:
    mkdir -p $HOME/out/SBI
    mv /output/sbi/${basename}*-train.sbi* $HOME/out/SBI/ || true
    mv /output/sbi/${basename}*-validation.sbi* $HOME/out/SBI/ || true
    mv /output/sbi/${basename}*-test.sbi* $HOME/out/SBI/  || true
    ls -lrt $HOME/out/SBI/

}
